File: src/c47/distributions/uniform.c
// SPDX-License-Identifier: GPL-3.0-only
// SPDX-FileCopyrightText: Copyright The WP43 and C47 Authors

/********************************************//**
 * \file uniform.c
 ***********************************************/

#include "c47.h"

static bool_t checkParamUniform(real_t *x, real_t *low, real_t *high, real_t *range, int *cmp, uint16_t discrete) {
  if (!saveLastX())
    return false;

  if (!getRegisterAsReal(REGISTER_X, x) || !getRegisterAsReal(REGISTER_M, low) || !getRegisterAsReal(REGISTER_N, high))
    goto err;
  if (realIsSpecial(x) || realIsSpecial(low) || realIsSpecial(high)) {
    displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_X);
    #if (EXTRA_INFO_ON_CALC_ERROR == 1)
      moreInfoOnError("In function checkParamUniform:", "given non-number inputs", NULL, NULL);
    #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
    goto err;
  }
  if (discrete) {
    if (!realIsAnInteger(low)) {
      displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_M);
      #if (EXTRA_INFO_ON_CALC_ERROR == 1)
        moreInfoOnError("In function checkParamUniform:", "given non-integer lower limit", NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      goto err;
    }
    if (!realIsAnInteger(high)) {
      displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_N);
      #if (EXTRA_INFO_ON_CALC_ERROR == 1)
        moreInfoOnError("In function checkParamUniform:", "given non-integer upper limit", NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      goto err;
    }
  }

  if (realCompareGreaterThan(low, high)) {
    const real_t t = *low;
    *low = *high;
    *high = t;
  }
  if (range != NULL) {
    realSubtract(high, low, range, &ctxtReal39);
    if (discrete)
      realAdd(range, const_1, range, &ctxtReal39);
  }
  if (cmp != NULL)
    *cmp = realCompareLessThan(x, low) ? -1 : realCompareGreaterThan(x, high) ? 1 : 0;
  return true;

err:
  if(getSystemFlag(FLAG_SPCRES)) {
    convertRealToResultRegister(const_NaN, REGISTER_X, amNone);
  }
  return false;
}

void fnUniformP(uint16_t discrete) {
  real_t x, l, h, range, *res = &x;
  int cmp;

  if (checkParamUniform(&x, &l, &h, &range, &cmp, discrete)) {
    if (cmp == 0) {
      if (realIsZero(&range))
        res = const_1;
      else if (discrete && !realIsAnInteger(&x))
        res = const_0;
      else
        realDivide(const_1, &range, &x, &ctxtReal39);
    } else
      res = const_0;
    convertRealToResultRegister(res, REGISTER_X, amNone);
    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }
}

void fnUniformL(uint16_t discrete) {
  real_t x, low, h, range, *res = &x;
  int cmp;

  if (checkParamUniform(&x, &low, &h, &range, &cmp, discrete)) {
    if (cmp < 0) {
      res = const_0;
    } else if (cmp > 0 || realIsZero(&range)) {
      res = const_1;
    } else {
      realSubtract(&x, &low, &h, &ctxtReal39);
      if (discrete) {
        realToIntegralValue(&h, &low, DEC_ROUND_FLOOR, &ctxtReal39);
        realAdd(&low, const_1, &h, &ctxtReal39);
      }
      realDivide(&h, &range, &x, &ctxtReal39);
    }
    convertRealToResultRegister(res, REGISTER_X, amNone);
    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }
}

void fnUniformU(uint16_t discrete) {
  real_t x, l, high, range, *res = &x;
  int cmp;

  if (checkParamUniform(&x, &l, &high, &range, &cmp, discrete)) {
    if (cmp < 0 || realIsZero(&range)) {
      res = const_1;
    } else if (cmp > 0) {
      res = const_0;
    } else {
      realSubtract(&high, &x, &l, &ctxtReal39);
      if (discrete) {
        realToIntegralValue(&l, &high, DEC_ROUND_CEILING, &ctxtReal39);
        realAdd(&high, const_1, &l, &ctxtReal39);
      }
      realDivide(&l, &range, &x, &ctxtReal39);
    }
    convertRealToResultRegister(res, REGISTER_X, amNone);
    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }
}

void fnUniformI(uint16_t discrete) {
  real_t x, low, high, res;

  if (checkParamUniform(&x, &low, &high, NULL, NULL, discrete)) {
    if (realCompareLessEqual(&x, const_0) || realCompareGreaterEqual(&x, const_1)) {
      displayDomainErrorMessage(ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, ERR_REGISTER_LINE, REGISTER_X);
      #if (EXTRA_INFO_ON_CALC_ERROR == 1)
        moreInfoOnError("In function fnUniformI:", "the argument must be 0 < x < 1", NULL, NULL);
      #endif // (EXTRA_INFO_ON_CALC_ERROR == 1)
      if(getSystemFlag(FLAG_SPCRES)) {
        convertRealToResultRegister(const_NaN, REGISTER_X, amNone);
      }
      return;
    }
    linpol(&low, &high, &x, &res);
    convertRealToResultRegister(&res, REGISTER_X, amNone);
    adjustResult(REGISTER_X, false, false, REGISTER_X, -1, -1);
  }
}

File: zig_src/frontier/distributions/uniform.zig
// SPDX-License-Identifier: GPL-3.0-only
//
// Zig owner for the (continuous and discrete) Uniform distribution commands,
// porting `src/c47/distributions/uniform.c`. The four fn* entry points each take
// a `discrete` flag (0 = continuous, non-zero = discrete) and are dispatched
// from the items.c function table via frontier.zig. All decNumber/register
// access goes through frontier_distribution_runtime. The testSuite has no
// Uniform dispatch entry, so these are verified by the distributions parity
// harness.

const dr = @import("../frontier_distribution_runtime.zig");
const real_t = dr.real_t;

// Reads X, low (REGISTER_M) and high (REGISTER_N); validates and, when
// requested, computes the range and the position of X relative to [low, high].
fn checkParamUniform(x: *real_t, low: *real_t, high: *real_t, range: ?*real_t, cmp: ?*c_int, discrete: u16) linksection(dr.code_section) bool {
    const ctx = &dr.ctxtReal39;
    if (!dr.saveLastX()) {
        return false;
    }

    if (!dr.getRegisterAsReal(dr.REGISTER_X, x) or !dr.getRegisterAsReal(dr.REGISTER_M, low) or !dr.getRegisterAsReal(dr.REGISTER_N, high)) {
        dr.specialResultNaN();
        return false;
    }
    if (dr.realIsSpecial(x) or dr.realIsSpecial(low) or dr.realIsSpecial(high)) {
        dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
        dr.moreInfoOnError("In function checkParamUniform:", "given non-number inputs", null, null);
        dr.specialResultNaN();
        return false;
    }
    if (discrete != 0) {
        if (!dr.isAnInteger(low)) {
            dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_M);
            dr.moreInfoOnError("In function checkParamUniform:", "given non-integer lower limit", null, null);
            dr.specialResultNaN();
            return false;
        }
        if (!dr.isAnInteger(high)) {
            dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_N);
            dr.moreInfoOnError("In function checkParamUniform:", "given non-integer upper limit", null, null);
            dr.specialResultNaN();
            return false;
        }
    }

    if (dr.realGreaterThan(low, high)) {
        const tmp = low.*;
        low.* = high.*;
        high.* = tmp;
    }
    if (range) |r| {
        dr.realSubtract(high, low, r, ctx);
        if (discrete != 0) {
            dr.realAdd(r, dr.const1(), r, ctx);
        }
    }
    if (cmp) |c| {
        c.* = if (dr.realLessThan(x, low)) @as(c_int, -1) else if (dr.realGreaterThan(x, high)) @as(c_int, 1) else @as(c_int, 0);
    }
    return true;
}

pub fn uniformP(discrete: u16) linksection(dr.code_section) void {
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var l: real_t = undefined;
    var h: real_t = undefined;
    var range: real_t = undefined;
    var cmp: c_int = undefined;
    var res: *const real_t = &x;

    if (checkParamUniform(&x, &l, &h, &range, &cmp, discrete)) {
        if (cmp == 0) {
            if (dr.realIsZero(&range)) {
                res = dr.const1();
            } else if (discrete != 0 and !dr.isAnInteger(&x)) {
                res = dr.const0();
            } else {
                dr.realDivide(dr.const1(), &range, &x, ctx);
            }
        } else {
            res = dr.const0();
        }
        dr.storeResult(res);
    }
}

pub fn uniformL(discrete: u16) linksection(dr.code_section) void {
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var low: real_t = undefined;
    var h: real_t = undefined;
    var range: real_t = undefined;
    var cmp: c_int = undefined;
    var res: *const real_t = &x;

    if (checkParamUniform(&x, &low, &h, &range, &cmp, discrete)) {
        if (cmp < 0) {
            res = dr.const0();
        } else if (cmp > 0 or dr.realIsZero(&range)) {
            res = dr.const1();
        } else {
            if (discrete != 0) {
                dr.toIntegralValue(&x, &h, dr.DEC_ROUND_FLOOR, ctx);
                dr.realAdd(&h, dr.const1(), &x, ctx);
            }
            dr.realSubtract(&x, &low, &h, ctx);
            dr.realDivide(&h, &range, &x, ctx);
        }
        dr.storeResult(res);
    }
}

pub fn uniformU(discrete: u16) linksection(dr.code_section) void {
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var l: real_t = undefined;
    var high: real_t = undefined;
    var range: real_t = undefined;
    var cmp: c_int = undefined;
    var res: *const real_t = &x;

    if (checkParamUniform(&x, &l, &high, &range, &cmp, discrete)) {
        if (cmp < 0 or dr.realIsZero(&range)) {
            res = dr.const1();
        } else if (cmp > 0) {
            res = dr.const0();
        } else {
            if (discrete != 0) {
                dr.toIntegralValue(&x, &l, dr.DEC_ROUND_CEILING, ctx);
                dr.realSubtract(&l, dr.const1(), &x, ctx);
            }
            dr.realSubtract(&high, &x, &l, ctx);
            dr.realDivide(&l, &range, &x, ctx);
        }
        dr.storeResult(res);
    }
}

pub fn uniformI(discrete: u16) linksection(dr.code_section) void {
    const ctx = &dr.ctxtReal39;
    var x: real_t = undefined;
    var low: real_t = undefined;
    var high: real_t = undefined;
    var t: real_t = undefined;

    if (checkParamUniform(&x, &low, &high, null, null, discrete)) {
        if (dr.realLessThan(&x, dr.const0()) or dr.realGreaterThan(&x, dr.const1())) {
            dr.displayDomainErrorMessage(dr.ERROR_ARG_EXCEEDS_FUNCTION_DOMAIN, dr.ERR_REGISTER_LINE, dr.REGISTER_X);
            dr.moreInfoOnError("In function fnUniformI:", "the argument must be 0 < x < 1", null, null);
            dr.specialResultNaN();
            return;
        }
        // Adjust an exact unity input to not go out of range.
        if (dr.realEqual(&x, dr.const1())) {
            dr.realCopy(&high, &x);
        } else {
            dr.realAdd(&high, dr.const1(), &high, ctx);
            dr.linearInterpolate(&low, &high, &x, &t);
            dr.toIntegralValue(&t, &x, dr.DEC_ROUND_FLOOR, ctx);
        }
        dr.storeResult(&x);
    }
}
